test_runner/src/main/kotlin/ftl/Main.kt

```diff
@@ -49,7 +49,12 @@ class Main : Runnable {
             // BugSnag opens a non-daemon thread which will keep the JVM process alive.
             // Flank must invoke exitProcess to exit cleanly.
             // https://github.com/bugsnag/bugsnag-java/issues/151
-            exitProcess(CommandLine(Main()).execute(*args))
+            try {
```

Comment:
Let's document in the code why we need `try/catch` since this isn't obvious.

Reply:
Also is there any way for us to add a JUnit test that will fail if we delete try/catch? I'm worried about "simplifying" this logic in the future and the test suite will still pass.

Reply:
To test, create a "malicious" command that starts a non-daemon thread, and then throws an Error.

The only issue is that this test will hang if the try/catch is removed. Instead of hanging, you could run the test itself in a separate thread, with a timer that ensures that the test completes within some timespan or else the test fails.

Something like this?

```java
import org.junit.jupiter.api.Test;
import picocli.CommandLine;
import picocli.CommandLine.Command;
import picocli.CommandLine.Unmatched;

import java.util.List;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.atomic.AtomicBoolean;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class TimeoutTest {
    @Command(name = "whosbad")
    static class Malicious implements Runnable {
        static CountDownLatch throwingError = new CountDownLatch(1);
        @Unmatched
        List<String> ignoredArgs;

        @Override
        public void run() {
            CountDownLatch isThreadRunning = new CountDownLatch(1);
            Thread forever = new Thread("forever") {
                public void run() {
                    try {
                        isThreadRunning.countDown();
                        Thread.sleep(Long.MAX_VALUE);
                    } catch (InterruptedException ignored) { }
                }
            };
            forever.setDaemon(false);
            forever.start();
            try {
                isThreadRunning.await();
            } catch (InterruptedException ignored) { }

            throwingError.countDown();
            throw new Error("Killing the calling thread...");
        }
    }

    static Integer exitedStatus = null;
    static void simulatedSystemExit(int status) {
        exitedStatus = status;
    }

    static class MyApp {
        public static void main(String... args) {
            try {
                simulatedSystemExit(new CommandLine(new Malicious()).execute(args));
            } catch (Throwable t) {
                t.printStackTrace();
                simulatedSystemExit(CommandLine.ExitCode.SOFTWARE);
            }
        }
    }

    @Test
    public void testSafeAgainstMaliciousActors() throws Exception {
        AtomicBoolean completed = new AtomicBoolean(false);
        Thread simulatedMain = new Thread("simulated-main") {
            public void run() {
                MyApp.main("some", "args");
                completed.set(true);
            }
        };
        simulatedMain.start();

        Malicious.throwingError.await();
        long TIMEOUT = 3 * 1000L;
        simulatedMain.join(TIMEOUT); // in this test, this returns immediately because the simulated-main thread was killed by the Error

        assertTrue(completed.get(), "Our simulated main thread should have completed but instead it hung...");
        assertEquals(CommandLine.ExitCode.SOFTWARE, exitedStatus);
    }
}
```

Reply:
That's awesome. Thanks for all your help with this issue!

Reply:
Updated to use a separate process instead of the same JVM as the test.